pkg/logging/journald_core.go

```diff
@@ -0,0 +1,122 @@
+	zapcore.DPanicLevel: journal.PriCrit,
+}
+
+// NewJournaldCore returns a zapcore.Core that sends log entries to systemd-journald and
+// uses the given identifier as a prefix for structured logging context that is sent as journal fields.
+// loggingFatal is set to true when fatals are logged so that other non-fatals can be swallowed.
+// You can find an example usage under Logging.loggingFatal.
+func NewJournaldCore(identifier string, loggingFatal *atomic.Bool, enab zapcore.LevelEnabler) zapcore.Core {
+	return &journaldCore{
+		LevelEnabler: enab,
+		identifier:   identifier,
+		loggingFatal: loggingFatal,
+	}
+}
+
+type journaldCore struct {
+	zapcore.LevelEnabler
+	context      []zapcore.Field
+	identifier   string
+	loggingFatal *atomic.Bool
+}
+
+func (j *journaldCore) Check(ent zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if j.Enabled(ent.Level) {
+		return ce.AddCore(ent, j)
+	}
+
+	return ce
+}
+
+func (j *journaldCore) Sync() error {
+	return nil
+}
+
+func (j *journaldCore) With(fields []zapcore.Field) zapcore.Core {
+	return &journaldCore{
+		LevelEnabler: j.LevelEnabler,
+		context:      append(j.context[:len(j.context):len(j.context)], fields...),
+		identifier:   j.identifier,
+		loggingFatal: j.loggingFatal,
+	}
+}
+
+// Write sends the log entry to systemd-journald.
+//
+// If fatal entries are logged, Write sleeps for a short time after the message has been written,
+// before zap then exits the program,
+// as systemd may not correctly attribute all or part of the log message due to a race condition,
+// and therefore it would be missing when checking logs via journalctl -u icingadb:
+// https://github.com/systemd/systemd/issues/2913
+//
+// Without sleep, we observed the following exit behavior when starting Icinga DB via systemd:
+//
+// • All or part of the log message is lost,
+// regardless of whether it is logged to stderr or to systemd-journald.
+//
+// • When writing to stderr after also writing to systemd-journald.
+// the message seems to be attributed correctly,
+// which is good for panics not under our control as they print to stderr.
+// For messages under our control,
+// this is actually not an option,
+// as we only want to log to systemd-journald.
+//
+// • Sleep before exiting helps systemd to attribute the message correctly,
+// regardless of whether it is written to systemd-journald or stderr.
+func (j *journaldCore) Write(ent zapcore.Entry, fields []zapcore.Field) error {
+	all := make([]zapcore.Field, 0, len(fields)+len(j.context))
+	all = append(all, j.context...)
+	all = append(all, fields...)
+
+	if ent.Level == zapcore.FatalLevel {
+		// Sleep between fatal log and os.Exit() to overcome a race condition that
+		// results in systemd no longer attributing messages from processes
+		// that have exited to their cgroup.
+		// See: https://github.com/systemd/systemd/issues/2913.
+		defer time.Sleep(time.Millisecond * 1100)
+
+		j.loggingFatal.Store(true)
```

Comment:
What do you think of additionally having it cancel a root context to also tear down as much as possible and not only hide further log messages?

Reply:
Since we wait for more than a second, why not.